src/System.Net.Http/src/System/Net/Http/Managed/AuthenticateAndRedirectHandler.cs

```diff
@@ -0,0 +1,231 @@
+                {
+                    location = new Uri(request.RequestUri, location);
+                }
+
+                // Disallow automatic redirection from secure to non-secure schemes
+                bool allowed =
+                    (HttpUtilities.IsSupportedNonSecureScheme(request.RequestUri.Scheme) && HttpUtilities.IsSupportedScheme(location.Scheme)) ||
+                    (HttpUtilities.IsSupportedSecureScheme(request.RequestUri.Scheme) && HttpUtilities.IsSupportedSecureScheme(location.Scheme));
+                if (!allowed)
+                {
+                    break;
+                }
+
+                redirectCount++;
+                if (redirectCount > _maxAutomaticRedirections)
+                {
+                    throw new HttpRequestException(SR.net_http_max_redirects);
+                }
+
+                // Set up for the automatic redirect
+                request.RequestUri = location;
+
+                if (RequestRequiresForceGet(response.StatusCode, request.Method))
+                {
+                    request.Method = HttpMethod.Get;
+                    request.Content = null;
+                }
+
+                // Do the redirect.
+                response.Dispose();
+            }
+
+            return response;
+        }
+
+        private bool RequestNeedsRedirect(HttpResponseMessage response)
+        {
+            // Return if redirect is not requested.
+            if (!_allowRedirect)
+                return false;
+
+            bool needRedirect = false;
+            switch (response.StatusCode)
+            {
+                case HttpStatusCode.Moved:
+                case HttpStatusCode.Found:
+                case HttpStatusCode.SeeOther:
+                case HttpStatusCode.TemporaryRedirect:
+                    needRedirect = true;
+                    break;
+
+                case HttpStatusCode.MultipleChoices:
+                    needRedirect = response.Headers.Location != null; // Don't redirect if no Location specified
+                    break;
+            }
+
+            return needRedirect;
+        }
+
+        private static bool RequestRequiresForceGet(HttpStatusCode statusCode, HttpMethod requestMethod)
+        {
+            if (statusCode == HttpStatusCode.Moved ||
+                statusCode == HttpStatusCode.Found ||
+                statusCode == HttpStatusCode.SeeOther ||
+                statusCode == HttpStatusCode.MultipleChoices)
+            {
+                return requestMethod == HttpMethod.Post;
+            }
+
+            return false;
+        }
+
+        private static AuthenticationHeaderValue GetSupportedAuthScheme(HttpHeaderValueCollection<AuthenticationHeaderValue> authenticateValues)
+        {
+            AuthenticationHeaderValue basicAuthenticationHeaderValue = null;
+
+            // Only Digest and Basic auth supported, ignore others.
+            foreach (AuthenticationHeaderValue ahv in authenticateValues)
```

Comment:
This is still using header order rather than giving Digest higher priority.

Reply:
how? I'm iterating through the values, and checking if digest is present, if it's present, then it is returning the digest scheme, otherwise it is returning the basic scheme.

Reply:
Nevermind, I see that you're not returning immediately for basic. That's an odd pattern. I guess it saves a loop compared to Contains.